dialer.go

```diff
@@ -45,26 +47,24 @@ func Dialer(opts *DialerOpts) DialFN {
 		opts.BufferSize = defaultBufferSize
 	}
 	d := &dialer{
-		dial:       opts.Dial,
-		bufferSize: opts.BufferSize,
-		poolSize:   opts.PoolSize,
+		DialerOpts: *opts,
 		pool:       make(map[string]map[int]*connAndSession)}
 	return d.Dial
 }
 
-func (d *dialer) Dial(network, addr string) (net.Conn, error) {
+func (d *dialer) Dial(ctx context.Context, network, addr string) (net.Conn, error) {
 	d.mx.Lock()
 	defer d.mx.Unlock()
```

Comment:
I do wonder if finer grained locking could help here, @oxtoacart. In particular it seems like not locking the connection creation itself could be a win. I've just non-scientifically noticed that first run seems to take a moment, and I wonder if this could be why.

Reply:
Although I suppose after that first KCP connection is established the future ones should be super quick...could always profile it but maybe not worth it..

Reply:
Yeah, the first one will take a little while and subsequent dial attempts will queue behind it, but they have to, otherwise we'll end up with multiple KCP connections.